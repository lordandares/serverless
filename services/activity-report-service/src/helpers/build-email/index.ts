import * as BPromise from 'bluebird'
import {
  filter,
  find,
  flow,
  get,
  getOr,
  isEmpty,
  reject,
  toUpper,
  trim,
} from 'lodash/fp'
import * as moment from 'moment-timezone'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

import { Location } from '../../handlers/email-processor'
import { getLocationReportLinks } from '../get-location-report-links'
import { UserDetails } from '../get-user-details'

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(true, 'options.formSubmissions'),
)

const getNoActivityOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(false, 'options.skipNoActivity'),
)

export interface BuildEmail extends UserDetails {
  locations: Location[]
}

export async function buildEmail(options: BuildEmail) {
  try {
    const {
      applicationName,
      email,
      firstName,
      locations,
      preferences,
      timezone,
    } = options

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-location-reports'
      : 'daily-location-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment()
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)
    const skipNoActivity = getNoActivityOption(preferences)

    const links = await BPromise.map(locations, location =>
      getLocationReportLinks({
        location,
        withFormSubmissions,
      }),
    )

    const activeLocations = filter('link')(links)
    const inactiveLocations = reject('link')(links)

    const hasNoLocations =
      isEmpty(activeLocations) && isEmpty(inactiveLocations)

    if (
      isEmpty(activeLocations) &&
      !isEmpty(inactiveLocations) && skipNoActivity
    ) {
      console.info('skipping activity report', {
        activeLocations: activeLocations.length,
        inactiveLocations: inactiveLocations.length,
        skipNoActivity,
      })
      return false
    }

    const name = trim(firstName)

    const data = JSON.stringify({
      activeLocations,
      applicationName,
      date: dateText,
      hasNoLocations,
      inactiveLocations,
      isProduction,
      name,
      sendInactive: !skipNoActivity,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, options })
    return false
  }
}
