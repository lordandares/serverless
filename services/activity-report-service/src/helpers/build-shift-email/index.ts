import * as BPromise from 'bluebird'
import {
  filter,
  isEmpty,
  reject,
  sortBy,
  toUpper,
  trim,
  getOr,
  flow,
  find,
  get,
} from 'lodash/fp'
import { orderBy } from 'lodash'
import * as moment from 'moment-timezone'

import { getShiftReportLinks } from '../get-shift-report-links'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-shift-report' }),
  getOr(true, 'options.formSubmissions'),
)

export async function buildShiftEmail({ userDetails, datetime }) {
  try {
    const {
      applicationName,
      email,
      firstName,
      preferences,
      timezone,
      shifts,
    } = userDetails

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-shift-reports'
      : 'daily-shift-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment(datetime)
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)

    const shiftLinks = await BPromise.map(shifts, shift =>
      getShiftReportLinks({
        datetime,
        shift,
        withFormSubmissions,
      }),
    )

    const activeShifts = orderBy(
      filter('link')(shiftLinks),
      ['lastName', 'timestamp'],
      ['asc'],
    )

    const hasNoShifts = isEmpty(activeShifts)
    const name = trim(firstName)

    const data = JSON.stringify({
      activeShifts,
      applicationName,
      date: dateText,
      hasNoShifts,
      isProduction,
      name,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, userDetails })
    return false
  }
}
