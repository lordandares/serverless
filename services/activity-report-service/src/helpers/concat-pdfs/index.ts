import { each } from 'lodash/fp'
import * as uuid from 'node-uuid'

import * as hummus from '@lighthouse/hummus-lambda'

export async function concatPdfs(filePaths, directoryPath) {
  const fileName = 'output.pdf'
  const filePath = `${directoryPath}/${fileName}`

  const pdfWriter = hummus.createWriter(filePath)
  each(path => appendPDFPageFromPDFWithAnnotations(pdfWriter, path), filePaths)
  pdfWriter.end()

  return filePath
}

/* istanbul ignore next */
function appendPDFPageFromPDFWithAnnotations(pdfWriter, sourcePDFPath) {
  const cpyCxt = pdfWriter.createPDFCopyingContext(sourcePDFPath)
  const cpyCxtParser = cpyCxt.getSourceDocumentParser()

  for (let i = 0; i < cpyCxtParser.getPagesCount(); ++i) {
    const pageDictionary = cpyCxtParser.parsePageDictionary(i)

    if (!pageDictionary.exists('Annots')) {
      // no annotation. append as is
      cpyCxt.appendPDFPageFromPDF(i)
    } else {
      // this let here will save any reffed objects from the copied annotations object.
      // they will be written after the page copy writing as to not to disturb the
      // page object writing itself.
      let reffedObjects

      pdfWriter.getEvents().once('OnPageWrite', params => {
        // using the page write event, write the new annotations. just copy the object
        // as is, saving any referenced objects for future writes
        params.pageDictionaryContext.writeKey('Annots')
        reffedObjects = cpyCxt.copyDirectObjectWithDeepCopy(
          pageDictionary.queryObject('Annots'),
        )
      })
      // write page. this will trigger the event
      cpyCxt.appendPDFPageFromPDF(i)
      // now write the reffed object (should be populated cause onPageWrite was written)
      // note that some or all annotations may be embedded, in which case this array
      // wont hold all annotation objects
      if (reffedObjects && reffedObjects.length > 0) {
        cpyCxt.copyNewObjectsForDirectObject(reffedObjects)
      }
    }
  }
}
