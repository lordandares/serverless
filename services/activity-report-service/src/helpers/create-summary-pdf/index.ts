import * as AWS from 'aws-sdk'
import { getOr } from 'lodash/fp'

import getPdfBuffer from '../get-pdf-buffer'
import * as s3 from '../s3'

interface Env {
  env: {
    S3_BUCKET_UPLOADS: string
    AWS_REGION: string
    S3_BASE_URL: string
    CLOUDINARY_BASE_URL: string
  }
}

declare var process: Env

export function buildConfig(data) {
  const { application } = data

  // pdf options
  const pageSize = getPageSize(process.env.AWS_REGION)
  const logoUrl = getOr(undefined, 'theme.logos.pdf', application)
  const pdfOptions = { logoUrl, pageSize }

  // custom application/dar settings
  const maxScans = getOr(undefined, 'settings.darMaxScans', application)
  const showAlternate = getOr(false, 'flags.darAlternateSummary', application)

  const settings = {
    awsS3BaseUrl: process.env.S3_BASE_URL,
    cloudinaryBaseUrl: process.env.CLOUDINARY_BASE_URL,
  }

  return {
    data: {
      ...data,
      maxScans,
      settings,
      showAlternate,
    },
    pdfOptions,
  }
}

export async function createSummaryPdf(pdfDefinition, s3FullPath) {
  const s3Client = new AWS.S3()
  const s3Bucket = process.env.S3_BUCKET_UPLOADS

  const buffer = await getPdfBuffer(pdfDefinition)
  await s3.put(s3Client, {
    buffer,
    bucket: s3Bucket,
    key: s3FullPath,
  })
}

// NOTE there's an issue with typescript compiled code and the branch coverage
// of istanbul. Disabling until finding an appropriate solution
// See: https://www.npmjs.com/package/ts-es5-istanbul-coverage

/* istanbul ignore next */
function getPageSize(region: string = '') {
  return region.startsWith('us-') ? 'LETTER' : 'A4'
}
