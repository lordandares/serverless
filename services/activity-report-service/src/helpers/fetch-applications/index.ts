import { mongo } from '@lighthouse/serverless-common'

export async function fetchApplications() {
  const collection = await mongo.getCollection('applications')

  return collection.find({}).toArray()
}
