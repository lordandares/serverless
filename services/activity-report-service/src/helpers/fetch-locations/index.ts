import { mongo } from '@lighthouse/serverless-common'
import { attempt, isError } from 'lodash/fp'

export async function fetchLocations(application: string) {
  const applicationId = attempt(() => new mongo.ObjectId(application))

  if (!application || isError(applicationId)) {
    throw new Error(
      'fetchLocations: Missing required environment variables / params',
    )
  }

  const collection = await mongo.getCollection('locations')

  return collection
    .find({
      application: applicationId,
      activityReportTriggerTime: {
        $exists: true,
      },
      deleted: { $ne: true },
      timezone: { $exists: true },
    })
    .toArray()
}
