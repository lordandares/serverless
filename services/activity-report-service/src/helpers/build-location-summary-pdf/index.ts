import { buildActivityPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildLocationSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildActivityPdf(pdfOptions, summaryData)
  const s3FullPath = getLocationSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getLocationSummaryPdfPath({ location }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: locationId } = location
  const s3PathName = `activity-reports/${locationId}/${year}/${month}`
  const s3FileName = `${year}-${month}-${day}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
