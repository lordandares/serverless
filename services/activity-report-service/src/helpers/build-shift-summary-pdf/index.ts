import { buildShiftPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildShiftSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildShiftPdf(pdfOptions, summaryData)
  const s3FullPath = getShiftSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getShiftSummaryPdfPath({ applicationId, shift }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: shiftId } = shift
  const s3PathName = `activity-reports/shifts/${applicationId}/${year}/${month}/${day}`
  const s3FileName = `${year}-${month}-${day}-shift-${shiftId}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
