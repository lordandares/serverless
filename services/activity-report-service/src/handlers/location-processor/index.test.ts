jest.mock('../../helpers/fetch-applications')
jest.mock('../../helpers/fetch-locations')

import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'
import * as BPromise from 'bluebird'

import {
  locationProcessor,
  parseApplicationLocations,
  processLocation,
} from './'

import { fetchApplications } from '../../helpers/fetch-applications'
import { fetchLocations } from '../../helpers/fetch-locations'

const TZ_AEST = 'Australia/Melbourne'
const TZ_NYC = 'America/New_York'

AWS.setSDKInstance(require('aws-sdk'))

describe('handlers:locationProcessor', () => {
  beforeAll(() => {
    jest.spyOn(BPromise, 'delay').mockImplementation(Promise.resolve)
  })

  afterEach(() => jest.clearAllMocks())

  describe('locationProcessor', () => {
    beforeEach(() =>
      AWS.mock('StepFunctions', 'startExecution', () => Promise.resolve()),
    )
    afterEach(() => AWS.restore('StepFunctions', 'startExecution'))

    it('should throw an error if the LOCATION_PROCESSOR_ARN environment variable is missing', () => {
      process.env.LOCATION_PROCESSOR_ARN = ''

      const event = {
        body: JSON.stringify({
          datetime: new Date(),
        }),
      }

      expect(locationProcessor(event)).rejects.toThrowError(
        /Missing required environment variables/,
      )
    })

    it('should reject with an error if applications can not be retrieved', () => {
      process.env.LOCATION_PROCESSOR_ARN =
        'arn:aws:lambda:us-east-1:account:function:test'
      fetchApplications.mockRejectedValue(new Error('not found error'))

      const event = {
        body: JSON.stringify({
          datetime: new Date(),
        }),
      }

      expect(locationProcessor(event)).rejects.toThrowError(/not found error/)
    })

    it('should successfully trigger a location activity report with current time', async () => {
      process.env.LOCATION_PROCESSOR_ARN =
        'arn:aws:lambda:us-east-1:account:function:test'

      // NOTE: outside of dst for AEST timezone
      MockDate.set('2018-11-09T04:45:00.000Z')

      const event = {}

      const applications = [
        {
          _id: '5c0efa65fc13ae1540000000',
          name: 'Lighthouse',
        },
      ]

      const locations = [
        {
          _id: '5c0efa65fc13ae1540000001',
          activityReportTriggerTime: {
            hours: 15, // in range
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
        {
          _id: '5c0efa65fc13ae1540000002',
          activityReportTriggerTime: {
            hours: 16, // out of range by hour
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
        {
          _id: '5c0efa65fc13ae1540000003',
          activityReportTriggerTime: {
            hours: 14, // out of range by hour
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
      ]

      fetchApplications.mockResolvedValue(applications)
      fetchLocations.mockResolvedValue(locations)

      const results = await locationProcessor(event)
      expect(results).toMatchSnapshot()
    })

    it('should successfully trigger a location activity report with custom time', async () => {
      process.env.LOCATION_PROCESSOR_ARN =
        'arn:aws:lambda:us-east-1:account:function:test'

      const event = {
        body: JSON.stringify({
          datetime: new Date('2018-11-09T05:45:00.000Z'),
        }),
      }

      const applications = [
        {
          _id: '5c0efa65fc13ae1540000000',
          name: 'Lighthouse',
        },
      ]

      const locations = [
        {
          _id: '5c0efa65fc13ae1540000001',
          activityReportTriggerTime: {
            hours: 15, // out of range by hour
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
        {
          _id: '5c0efa65fc13ae1540000002',
          activityReportTriggerTime: {
            hours: 16, // in range
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
        {
          _id: '5c0efa65fc13ae1540000003',
          activityReportTriggerTime: {
            hours: 14, // out of range by hour
            minutes: 0,
          },
          application: '5c0efa65fc13ae1540000000',
          timezone: TZ_AEST,
        },
      ]

      fetchApplications.mockResolvedValue(applications)
      fetchLocations.mockResolvedValue(locations)

      const results = await locationProcessor(event)
      expect(results).toMatchSnapshot()
    })
  })

  describe('parseApplicationLocations', () => {
    it('should return an empty array if no application locations are passed', () => {
      expect.assertions(1)

      const applicationLocations = []

      expect(
        parseApplicationLocations(applicationLocations, new Date()),
      ).toEqual([])
    })

    it('should return an empty array if application locations missing trigger time or timezone', () => {
      expect.assertions(1)

      const applicationLocations = [
        {
          _id: 1,
          activityReportTriggerTime: {
            hours: 15,
            minutes: 0,
          },
          application: 101,
        },
        {
          _id: 2,
          application: 101,
          timezone: TZ_AEST,
        },
      ]

      expect(
        parseApplicationLocations(applicationLocations, new Date()),
      ).toEqual([])
    })

    it('should process application locations', () => {
      expect.assertions(1)

      // NOTE: outside of dst for AEST timezone
      MockDate.set('2018-11-09T16:30:00.000Z')

      const applicationLocationArray = [
        {
          _id: 1,
          activityReportTriggerTime: {
            hours: 3, // in range
            minutes: 0,
          },
          application: 101,
          timezone: TZ_AEST,
        },
        {
          _id: 2,
          activityReportTriggerTime: {
            hours: 3, // in range
            minutes: 0,
          },
          application: 101,
          timezone: TZ_AEST,
        },
        {
          _id: 3,
          activityReportTriggerTime: {
            hours: 4, // out of range by one hour
            minutes: 0,
          },
          application: 101,
          timezone: TZ_AEST,
        },
        {
          _id: 4,
          activityReportTriggerTime: {
            hours: 2, // out of range by one hour
            minutes: 0,
          },
          application: 101,
          timezone: TZ_AEST,
        },
      ]

      const applicationLocations = [applicationLocationArray]
      const locationsToProcess = parseApplicationLocations(
        applicationLocations,
        new Date(),
      )
      expect(locationsToProcess).toMatchSnapshot()
      MockDate.reset()
    })

    describe('daylight saving time', () => {
      it('is handled when it starts in a timezone', () => {
        expect.assertions(3)

        // TZ_NYC clocks go forward Sunday, 10th March 2019, 2:00:00 am
        const applicationLocationArray = [
          {
            _id: 1,
            activityReportTriggerTime: {
              hours: 6,
              minutes: 0,
            },
            application: 101,
            timezone: TZ_NYC,
          },
        ]
        const applicationLocations = [applicationLocationArray]

        // Before clocks go forward
        MockDate.set('2019-03-09T10:00:00.000Z') // NYC 2019-03-09T05:00:00.000Z UTC -5
        const locationsToProcessBefore = parseApplicationLocations(
          applicationLocations,
          new Date(),
        )
        expect(locationsToProcessBefore.length).toEqual(0)

        // After clocks go forward
        MockDate.set('2019-03-11T10:00:00.000Z') // NYC 2019-03-09T06:00:00.000Z UTC -4
        const locationsToProcessAfter = parseApplicationLocations(
          applicationLocations,
          new Date(),
        )
        expect(locationsToProcessAfter.length).toEqual(1)
        expect(locationsToProcessAfter).toMatchSnapshot()
        MockDate.reset()
      })

      it('is handled when it ends in a timezone', () => {
        expect.assertions(3)

        // TZ_NYC clocks go back Sunday, 3 November 2019, 2:00:00 am
        const applicationLocationArray = [
          {
            _id: 1,
            activityReportTriggerTime: {
              hours: 6,
              minutes: 0,
            },
            application: 101,
            timezone: TZ_NYC,
          },
        ]
        const applicationLocations = [applicationLocationArray]

        // Before clocks go back
        MockDate.set('2019-11-02T10:00:00.000Z') // NYC 2019-11-02T06:00:00.000Z UTC -4
        const locationsToProcessBefore = parseApplicationLocations(
          applicationLocations,
          new Date(),
        )

        expect(locationsToProcessBefore.length).toEqual(1)
        expect(locationsToProcessBefore).toMatchSnapshot()

        // After clocks have gone back
        MockDate.set('2019-11-04T10:00:00.000Z') // NYC 2019-11-02T05:00:00.000Z UTC -5
        const locationsToProcessAfter = parseApplicationLocations(
          applicationLocations,
          new Date(),
        )
        expect(locationsToProcessAfter.length).toEqual(0)
        MockDate.reset()
      })
    })
  })

  describe('processLocation', () => {
    beforeEach(() => {
      process.env.MONGODB_SECRET_ID = 'lio/application/service/environment'

      AWS.mock(
        'StepFunctions',
        'startExecution',
        // NOTE: Just returning the params here,
        // as no need to test the aws-sdk
        (params, cb) => cb(null, params),
      )
    })

    afterEach(() => AWS.restore('StepFunctions', 'startExecution'))

    it('should error if LOCATION_PROCESSOR_ARN is missing', () => {
      process.env.LOCATION_PROCESSOR_ARN = ''

      expect(() => processLocation({})).toThrowError(
        /Missing LOCATION_PROCESSOR_ARN/,
      )
    })

    it('should process location', async () => {
      process.env.LOCATION_PROCESSOR_ARN =
        'arn:aws:lambda:us-east-1:account:function:test'
      const location = { _id: 1, application: 2 }

      const expectedResponse = {
        input: JSON.stringify(location),
        stateMachineArn: 'arn:aws:lambda:us-east-1:account:function:test',
      }

      const processed = await processLocation(location)

      expect(processed).toEqual(expectedResponse)
    })
  })
})
