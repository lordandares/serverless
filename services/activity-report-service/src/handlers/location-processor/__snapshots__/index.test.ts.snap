// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers:locationProcessor locationProcessor should successfully trigger a location activity report with current time 1`] = `
Object {
  "locationsProcessedCount": 1,
}
`;

exports[`handlers:locationProcessor locationProcessor should successfully trigger a location activity report with custom time 1`] = `
Object {
  "locationsProcessedCount": 1,
}
`;

exports[`handlers:locationProcessor parseApplicationLocations daylight saving time is handled when it ends in a timezone 1`] = `
Array [
  Object {
    "applicationId": 101,
    "end": "2019-11-02T09:59:59+00:00",
    "locationId": 1,
    "start": "2019-11-01T10:00:00+00:00",
  },
]
`;

exports[`handlers:locationProcessor parseApplicationLocations daylight saving time is handled when it starts in a timezone 1`] = `
Array [
  Object {
    "applicationId": 101,
    "end": "2019-03-11T09:59:59+00:00",
    "locationId": 1,
    "start": "2019-03-10T10:00:00+00:00",
  },
]
`;

exports[`handlers:locationProcessor parseApplicationLocations should process application locations 1`] = `
Array [
  Object {
    "applicationId": 101,
    "end": "2018-11-09T15:59:59+00:00",
    "locationId": 1,
    "start": "2018-11-08T16:00:00+00:00",
  },
  Object {
    "applicationId": 101,
    "end": "2018-11-09T15:59:59+00:00",
    "locationId": 2,
    "start": "2018-11-08T16:00:00+00:00",
  },
]
`;
